engine/crates/integration-tests/tests/federation/basic/variables.rs

```diff
@@ -44,42 +48,398 @@ fn lists() {
+            "string": "hello",
+            "int": 1,
+            "float": 1.0
+        }),
+    );
+    roundtrip_test(
+        "inputObject",
+        "InputObj!",
+        json!({
+            "string": "hello",
+            "int": 1,
+            "float": 1.0
+        }),
+    );
+    roundtrip_test(
+        "inputObject",
+        "InputObj!",
+        json!({
+            "string": null,
+        }),
+    );
+    roundtrip_test(
+        "inputObject",
+        "InputObj!",
+        json!({
+            "recursiveObject": {"string": null}
+        }),
+    );
+    roundtrip_test(
+        "inputObject",
+        "InputObj!",
+        json!({
+            "recursiveObject": {"recursiveObject": {"string": "hello"}}
+        }),
+    );
+    roundtrip_test(
+        "inputObject",
+        "InputObj!",
+        json!({
+            "recursiveObject": null
+        }),
+    );
+    roundtrip_test(
+        "inputObject",
+        "InputObj!",
+        json!({
+            "recursiveObjectList": null
+        }),
+    );
+    roundtrip_test(
+        "inputObject",
+        "InputObj!",
+        json!({
+            "recursiveObjectList": [{"string": "hello"}]
+        }),
+    );
+    roundtrip_test(
+        "inputObject",
+        "InputObj!",
+        json!({
+            "recursiveObjectList": [{"recursiveObject": {"string": "hello"}}]
+        }),
+    );
 }
 
 #[test]
-fn list_coercion() {
-    let query = "query($input: String!) { listOfListOfStrings(input: $input) }";
-    let input = json!("hello");
+fn test_default_values() {
+    let query = r#"query($input: String = "there") { listOfListOfStrings(input: $input) }"#;
+    let input = json!({"input": "hello"});
+    assert_eq!(
+        run_query(query, &input).into_data()["listOfListOfStrings"],
+        json!([["hello"]])
+    );
 
-    let response = runtime().block_on({
-        let input = input.clone();
-        async move {
-            let echo_mock = MockGraphQlServer::new(EchoSchema::default()).await;
+    let input = json!({});
+    assert_eq!(
+        run_query(query, &input).into_data()["listOfListOfStrings"],
+        json!([["there"]])
+    );
+}
 
-            let engine = Engine::build().with_schema("schema", &echo_mock).await.finish();
+#[test]
+fn list_coercion() {
+    let query = "query($input: [[String!]!]!) { listOfListOfStrings(input: $input) }";
+    let input = json!({"input": "hello"});
+    assert_eq!(
+        run_query(query, &input).into_data()["listOfListOfStrings"],
+        json!([["hello"]])
+    );
 
-            engine.execute(query).variables(json!({"input": input})).await
-        }
+    let query = "query($input: [String!]!) { listOfStrings(input: $input) }";
+    assert_eq!(run_query(query, &input).into_data()["listOfStrings"], json!(["hello"]));
+
+    let query = "query($input: InputObj!) { inputObject(input: $input) }";
+    let input = json!({
+        "input": {"annoyinglyOptionalStrings": "hello", "recursiveObjectList": {"id": "1"}}
```

Comment:
TIL single values coerce to a one-item list. I didn't know that part of the spec.